### PR TITLE
chore: Add Che update e2e test

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### What does this PR do?
Adds new PR check that tests Che update flow using `operator` installer.
The test does the following:
1. Deploys latest stable version of Che.
2. Creates a workspace, starts (waits for `RUNNING` state) and then stops it.
3. Updates Che to nightly version (the one under development).
4. Starts the workspace to check that nothing is wrong with old workspaces.
5. Downgrades Che back to the latest stable version.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/19013

### How to test this PR?
Run 
```sh
minikube start && minikube addons enable ingress
yarn test --coverage=false --forceExit --testRegex=test/e2e/e2e-upgrade.test.ts
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by flacatus — APPROVED** [review]

File: test/e2e/e2e-upgrade.test.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
+/*********************************************************************
+ * Copyright (c) 2020 Red Hat, Inc.
```
Comment (AndrienkoAleksandr) [inline test/e2e/e2e-upgrade.test.ts:2]: minor 2021

File: test/e2e/e2e-upgrade.test.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    it('Downgrade Che', async () => {
+      await runCommand(binChectl, ['server:update', '-y', `--version=${cheVersion}`, `-n ${NAMESPACE}`, '--telemetry=off'])
+      expect(await waitForCheServerImageTag(cheVersion, UPDATE_CHE_TIMEOUT_MS)).equal(true)
+    })
+
+    it('Check downgraded Che version', async () => {
+      expect(await waitForVersionInCheCR(cheVersion, CHE_VERSION_TIMEOUT_MS)).equal(true)
+    })
+  })
+
```
Comment (AndrienkoAleksandr) [inline test/e2e/e2e-upgrade.test.ts:154]: Maybe we should check that workspace starts after downgrade.... But up to you.

**Review by AndrienkoAleksandr — APPROVED** [review]

File: test/e2e/e2e-upgrade.test.ts (hunk 2 above)
Comment (mmorhun) [reply test/e2e/e2e-upgrade.test.ts:154]: The use-case in the tests, at least my thoughts about it, is that downgrade is used if a user have completed upgrade, but something stopped to work due to a bug, but the user need the feature, so a step back is taken temporally and right after upgrade... So no need in such check.
If a used needs specific version of Che, it is possible to install it via `version` flag, no need to downgrade.

File: test/e2e/util.ts [hunk 3]
```diff
@@ -85,6 +87,20 @@ export class E2eHelper {
     return workspaceStatus
   }
 
+  async waitWorkspaceStatus(status: string, timeoutMs: number): Promise<boolean> {
```
Comment (tolusha) [inline test/e2e/util.ts:90]: Usually `wait` methods return void or throw an error.

File: test/e2e/e2e-upgrade.test.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  if (exitCode !== 0) {
+    console.log(stderr)
+  }
+
+  expect(exitCode).equal(0)
+
+  return stdout
+}
+
+async function waitForVersionInCheCR(version: string, timeoutMs: number): Promise<boolean> {
```
Comment (tolusha) [inline test/e2e/e2e-upgrade.test.ts:52]: Usually wait methods return void or throw an error.
Should be moved to utility class.

File: test/e2e/e2e-upgrade.test.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    if (cheCR.status.cheVersion === version) {
+      return true
+    }
+    await helper.sleep(delayMs)
+    totalTimeMs += delayMs
+  }
+  return false
+}
+
+async function waitForCheServerImageTag(tag: string, timeoutMs: number): Promise<boolean> {
```
Comment (tolusha) [inline test/e2e/e2e-upgrade.test.ts:67]: Usually wait methods return void or throw an error.
Should be moved to utility class.

File: test/e2e/e2e-upgrade.test.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  return stdout
+}
+
+async function waitForVersionInCheCR(version: string, timeoutMs: number): Promise<boolean> {
+  const delayMs = 5 * 1000
+
+  let totalTimeMs = 0
+  while (totalTimeMs < timeoutMs) {
+    const cheCR = await kube.getCheCluster(NAMESPACE)
+    if (cheCR.status.cheVersion === version) {
```
Comment (tolusha) [inline test/e2e/e2e-upgrade.test.ts:58]:
```suggestion
    if (cheCR && cheCR.status && cheCR.status.cheVersion === version) {
```

File: test/e2e/e2e-upgrade.test.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    })
+
+    it('Check downgraded Che version', async () => {
+      expect(await waitForVersionInCheCR(cheVersion, CHE_VERSION_TIMEOUT_MS)).equal(true)
+    })
+  })
+
+})
+
+async function loginTest() {
```
Comment (tolusha) [inline test/e2e/e2e-upgrade.test.ts:157]: Should be moved to utility class.
May be `login`,  `runLoginCommand` or `loginIntoEclipseChe` ?

File: .github/workflows/e2e-minikube-operator-update.yml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#  Copyright (c) 2021 Red Hat, Inc.
+#    This program and the accompanying materials are made
+#    available under the terms of the Eclipse Public License 2.0
+#    which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+#  SPDX-License-Identifier: EPL-2.0
+#
+#  Contributors:
+#    Red Hat, Inc. - initial API and implementation
+name: Minikube E2E Che update
```
Comment (tolusha) [inline .github/workflows/e2e-minikube-operator-update.yml:11]:
```suggestion
name: Minikube
```

File: .github/workflows/e2e-minikube-operator-update.yml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+on: pull_request
+jobs:
+  minikube-e2e-update:
+    name: Operator installer update
```
Comment (tolusha) [inline .github/workflows/e2e-minikube-operator-update.yml:15]:
```suggestion
    name: e2e update tests
```

File: test/e2e/util.ts (hunk 3 above)
Comment (mmorhun) [reply test/e2e/util.ts:90]: The idea is to wait for an event and then decide what to do: fail or continue. But maybe it is not needed for tests...

File: test/e2e/e2e-upgrade.test.ts (hunk 4 above)
Comment (mmorhun) [reply test/e2e/e2e-upgrade.test.ts:52]: ok

File: test/e2e/e2e-upgrade.test.ts (hunk 5 above)
Comment (mmorhun) [reply test/e2e/e2e-upgrade.test.ts:67]: ok

**Review by tolusha — APPROVED** [review]

**Pull request merged**